os/lib/sha-256.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright (c) 2021, Uppsala universitet.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * This file is part of the Contiki operating system.
+ *
+ */
+
+/**
+ * \file
+ *         Platform-independent SHA-256 API.
+ * \author
+ *         Konrad Krentz <konrad.krentz@gmail.com>
+ */
+
+#ifndef SHA_256_H_
+#define SHA_256_H_
+
+#include "contiki.h"
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#define SHA_256_DIGEST_LENGTH 32
+#define SHA_256_BLOCK_SIZE 64
+
+#ifdef SHA_256_CONF
+#define SHA_256 SHA_256_CONF
+#else /* SHA_256_CONF */
+#define SHA_256 sha_256_driver
+#endif /* SHA_256_CONF */
+
+typedef union {
+  struct {
+    /* used by software implementation */
+    uint64_t count;
+    uint32_t state[SHA_256_DIGEST_LENGTH / sizeof(uint32_t)];
+    uint8_t buf[SHA_256_BLOCK_SIZE];
+  } soft;
+
+  struct {
+    /* used by CC2538 implementation */
+    uint32_t state[SHA_256_DIGEST_LENGTH / sizeof(uint32_t)];
+    size_t hashed_len;
+    size_t buf_len;
+    uint8_t buf[SHA_256_BLOCK_SIZE];
+    bool was_crypto_enabled;
+  } cc2538;
+} sha_256_state_t;
```

Comment:
This seems out-of-place being in the API of a platform-independent library in `os`. I haven't had time to do a proper review, but isn't it possible to avoid this? Are by any chance those fields needed by all hardware-implementations?

Reply:
Good proposal. Now, both of my drivers use the same context variables.

I was also thinking about using static variables inside each implementation instead of an externally managed state. This would simplify the API a lot, but it would no longer be possible to do `PROCESS_PAUSE` between `SHA_256.init`, `SHA_256.update`, and `SHA_256.finalize`.